From c6fc1c027633f5bc12ba1ca4a729efdafbcfe2c7 Mon Sep 17 00:00:00 2001
From: Stanislaw Kardach <skardach@marvell.com>
Date: Wed, 15 Apr 2020 18:15:44 +0200
Subject: [PATCH 0452/1921] octeontx2-af: fix implitit variable array

In LF teardown of SSO attached to NPA the array storing possible matches
for cleanup was declared as a variable array. This is forbidden in C90.
Therefore dynamically allocate this array.

Change-Id: I56be40dca0db5d5a70ce2743037a5059a44e2e78
Signed-off-by: Stanislaw Kardach <skardach@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/26947
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/rvu.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
index 0024e2fea567..bbdc6e4e9d4e 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
@@ -2068,13 +2068,15 @@ static void rvu_npa_lf_mapped_nix_lf_teardown(struct rvu *rvu, u16 pcifunc)
 
 static void rvu_npa_lf_mapped_sso_lf_teardown(struct rvu *rvu, u16 pcifunc)
 {
-	u16 pcifunc_arr[rvu->hw->total_pfs + rvu->hw->total_vfs];
+	u16 *pcifunc_arr;
 	u16 sso_pcifunc, match_cnt = 0;
 	struct rvu_block *sso_block;
 	struct rsrc_detach detach;
 	int blkaddr, lf;
 	u64 regval;
 
+	pcifunc_arr = kcalloc(rvu->hw->total_pfs + rvu->hw->total_vfs,
+			      sizeof(*pcifunc_arr), GFP_KERNEL);
 	blkaddr = rvu_get_blkaddr(rvu, BLKTYPE_SSO, 0);
 	if (blkaddr < 0)
 		return;
@@ -2102,6 +2104,7 @@ static void rvu_npa_lf_mapped_sso_lf_teardown(struct rvu *rvu, u16 pcifunc)
 
 	for (sso_pcifunc = 0; sso_pcifunc < match_cnt; sso_pcifunc++)
 		rvu_detach_rsrcs(rvu, &detach, pcifunc_arr[sso_pcifunc]);
+	kfree(pcifunc_arr);
 }
 
 static void rvu_blklf_teardown(struct rvu *rvu, u16 pcifunc, u8 blkaddr)
-- 
2.31.1

